### ISSUE TITLE
Memory model: The hard bits

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11375]
Memory model: The hard bits<br>
<br>
https://ift.tt/1hXUZnA<br>
<br>
<br>
<br>

<div><div><div>

<p>This chapter describes the details of OCaml relaxed memory model. The relaxed
memory model describes what values an OCaml program is allowed to witness when
reading a memory location. If you are interested in high-level parallel
programming in OCaml, please have a look at the parallel programming chapter
<a href="https://ocaml.org/manual/5.2/parallelism.html#c%3Aparallelism">9</a>.</p>
<p>This chapter is aimed at experts who would like to understand the details of
the OCaml memory model from a practitioner’s perspective. For a formal
definition of the OCaml memory model, its guarantees and the compilation to
hardware memory models, please have a look at the PLDI 2018 paper on
<a href="https://doi.org/10.1145/3192366.3192421">Bounding Data Races in Space and
Time</a>. The memory model presented in this chapter is an extension of the one
presented in the PLDI 2018 paper. This chapter also covers some pragmatic
aspects of the memory model that are not covered in the paper.</p>
<h2 id="s:why_relaxed_memory">
<span>1</span> Why weakly consistent memory?</h2>
<p>The simplest memory model that we could give to our programs is sequential
consistency. Under sequential consistency, the values observed by the program
can be explained through some interleaving of the operations from different
domains in the program. For example, consider the following program with two
domains <span><span>d1</span></span> and <span><span>d2</span></span> executing in parallel:</p>
<div>

<div>

<p><span>let</span> d1 a b =
  <span>let</span> r1 = !a * 2 <span>in</span>
  <span>let</span> r2 = !b <span>in</span>
  <span>let</span> r3 = !a * 2 <span>in</span>
  (r1, r2, r3)

<span>let</span> d2 b = b := 0

<span>let</span> main () =
  <span>let</span> a = <span>ref</span> 1 <span>in</span>
  <span>let</span> b = <span>ref</span> 1 <span>in</span>
  <span>let</span> h = Domain.spawn (<span>fun</span> _ -&gt;
    <span>let</span> r1, r2, r3 = d1 a b <span>in</span>
    Printf.printf <span>"r1 = %d, r2 = %d, r3 = %d\n"</span> r1 r2 r3)
  <span>in</span>
  d2 b;
  Domain.join h</p>
</div>

</div>
<p>The reference cells <span><span>a</span></span> and <span><span>b</span></span> are initially <span><span>1</span></span>. The user may observe <span><span>r1 = 2, r2 = 0, r3 = 2</span></span> if the write to <span><span>b</span></span> in <span><span>d2</span></span> occurred before the read of <span><span>b</span></span>
in <span><span>d1</span></span>. Here, the observed behaviour can be explained in terms of interleaving
of the operations from different domains.</p>
<p>Let us now assume that <span><span>a</span></span> and <span><span>b</span></span> are aliases of each other.</p>
<div>

<div>

<p><span>let</span> d1 a b =
  <span>let</span> r1 = !a * 2 <span>in</span>
  <span>let</span> r2 = !b <span>in</span>
  <span>let</span> r3 = !a * 2 <span>in</span>
  (r1, r2, r3)

<span>let</span> d2 b = b := 0

<span>let</span> main () =
  <span>let</span> ab = <span>ref</span> 1 <span>in</span>
  <span>let</span> h = Domain.spawn (<span>fun</span> _ -&gt;
    <span>let</span> r1, r2, r3 = d1 ab ab <span>in</span>
    <span>assert</span> (not (r1 = 2 &amp;&amp; r2 = 0 &amp;&amp; r3 = 2)))
  <span>in</span>
  d2 ab;
  Domain.join h</p>
</div>

</div>
<p>In the above program, the variables <span><span>ab</span></span>, <span><span>a</span></span> and <span><span>b</span></span> refer to the same
reference cell. One would expect that the assertion in the main function will
never fail. The reasoning is that if <span><span>r2</span></span> is <span><span>0</span></span>, then the write in <span><span>d2</span></span>
occurred before the read of <span><span>b</span></span> in <span><span>d1</span></span>. Given that <span><span>a</span></span> and <span><span>b</span></span> are aliases,
the second read of <span><span>a</span></span> in <span><span>d1</span></span> should also return <span><span>0</span></span>.</p>
<h3 id="ss:mm_comp_opt">
<span>1.1</span> Compiler optimisations</h3>
<p>Surprisingly, this assertion may fail in OCaml due to compiler optimisations.
The OCaml compiler observes the common sub-expression <span><span>!a * 2</span></span> in <span><span>d1</span></span> and
optimises the program to:</p>
<div>

<div>

<p><span>let</span> d1 a b =
  <span>let</span> r1 = !a * 2 <span>in</span>
  <span>let</span> r2 = !b <span>in</span>
  <span>let</span> r3 = r1 <span>in</span> 
  (r1, r2, r3)

<span>let</span> d2 b = b := 0

<span>let</span> main () =
  <span>let</span> ab = <span>ref</span> 1 <span>in</span>
  <span>let</span> h = Domain.spawn (<span>fun</span> _ -&gt;
    <span>let</span> r1, r2, r3 = d1 ab ab <span>in</span>
    <span>assert</span> (not (r1 = 2 &amp;&amp; r2 = 0 &amp;&amp; r3 = 2)))
  <span>in</span>
  d2 ab;
  Domain.join h</p>
</div>

</div>
<p>This optimisation is known as the common sub-expression elimination (CSE). Such
optimisations are valid and necessary for good performance, and do not change
the sequential meaning of the program. However, CSE breaks sequential
reasoning.</p>
<p>In the optimized program above, even if the write to <span><span>b</span></span> in <span><span>d2</span></span> occurs between
the first and the second reads in <span><span>d1</span></span>, the program will observe the value <span><span>2</span></span>
for <span><span>r3</span></span>, causing the assertion to fail. The observed behaviour cannot be
explained by interleaving of operations from different domains in the source
program. Thus, CSE optimization is said to be invalid under sequential
consistency.</p>
<p>One way to explain the observed behaviour is as if the operations performed on
a domain were reordered. For example, if the second and the third reads from
<span><span>d1</span></span> were reordered,</p>
<div>

<div>

<p><span>let</span> d1 a b =
  <span>let</span> r1 = !a * 2 <span>in</span>
  <span>let</span> r3 = !a * 2 <span>in</span>
  <span>let</span> r2 = !b <span>in</span>
  (r1, r2, r3)</p>
</div>

</div>
<p>then we can explain the observed behaviour <span><span>(2,0,2)</span></span> returned by
<span><span>d1</span></span>.</p>
<h3 id="ss:mm_hw_opt">
<span>1.2</span> Hardware optimisations</h3>
<p>The other source of reordering is by the hardware. Modern hardware
architectures have complex cache hierarchies with multiple levels of cache.
While cache coherence ensures that reads and writes to a single memory
location respect sequential consistency, the guarantees on programs that
operate on different memory locations are much weaker. Consider the following
program:</p>
<div>

<div>

<p><span>let</span> a = <span>ref</span> 0
<span>and</span> b = <span>ref</span> 0

<span>let</span> d1 () =
  a := 1;
  !b

<span>let</span> d2 () =
  b := 1;
  !a

<span>let</span> main () =
  <span>let</span> h = Domain.spawn d2 <span>in</span>
  <span>let</span> r1 = d1 () <span>in</span>
  <span>let</span> r2 = Domain.join h <span>in</span>
  <span>assert</span> (not (r1 = 0 &amp;&amp; r2 = 0))</p>
</div>

</div>
<p>Under sequential consistency, we would never expect the assertion to fail.
However, even on x86, which offers much stronger guarantees than ARM, the
writes performed at a CPU core are not immediately published to all of the
other cores. Since <span><span>a</span></span> and <span><span>b</span></span> are different memory locations, the reads of <span><span>a</span></span>
and <span><span>b</span></span> may both witness the initial values, leading to the assertion failure.</p>
<p>This behaviour can be explained if a load is allowed to be reordered before a
preceding store to a different memory location. This reordering can happen due
to the presence of in-core store-buffers on modern processors. Each core
effectively has a FIFO buffer of pending writes to avoid the need to block
while a write completes. The writes to <span><span>a</span></span> and <span><span>b</span></span> may be in the store-buffers
of cores <span><span>c1</span></span> and <span><span>c2</span></span> running the domains <span><span>d1</span></span> and <span><span>d2</span></span>, respectively. The
reads of <span><span>b</span></span> and <span><span>a</span></span> running on the cores <span><span>c1</span></span> and <span><span>c2</span></span>, respectively, will not
see the writes if the writes have not propagated from the buffers to the main
memory.</p>
<h2 id="s:drf_sc">
<span>2</span> Data race freedom implies sequential consistency</h2>
<p>The aim of the OCaml relaxed memory model is to precisely describe which orders
are preserved by the OCaml program. The compiler and the hardware are free to
optimize the program as long as they respect the ordering guarantees of the
memory model. While programming directly under the relaxed memory model is
difficult, the memory model also describes the conditions under which a program
will only exhibit sequentially consistent behaviours. This guarantee is known
as <em>data race freedom implies sequential consistency</em> (DRF-SC). In this
section, we shall describe this guarantee. In order to do this, we first need a
number of definitions.</p>
<h3 id="s:atomics">
<span>2.1</span> Memory locations</h3>
<p>OCaml classifies memory locations into <em>atomic</em> and <em>non-atomic</em>
locations. Reference cells, array fields and mutable record fields are
non-atomic memory locations. Immutable objects are non-atomic locations with an
initialising write but no further updates. Atomic memory locations are those
that are created using the <a href="https://ocaml.org/manual/5.2/../5.2/api/Atomic.html"><span>Atomic</span></a> module.</p>
<h3 id="s:happens_before">
<span>2.2</span> Happens-before relation</h3>
<p>Let us imagine that the OCaml programs are executed by an abstract machine that
executes one action at a time, arbitrarily picking one of the available domains
at each step. We classify actions into two: <em>inter-domain</em> and
<em>intra-domain</em>. An inter-domain action is one which can be observed and be
influenced by actions on other domains. There are several inter-domain actions:</p>
<ul>
<li>
Reads and writes of atomic and non-atomic locations.
</li>
<li>Spawn and join of domains.
</li>
<li>Operations on mutexes.
</li>
</ul>
<p>On the other hand, intra-domain actions can neither be observed nor influence
the execution of other domains. Examples include evaluating an arithmetic
expression, calling a function, etc. The memory model specification ignores
such intra-domain actions. In the sequel, we use the term action to indicate
inter-domain actions.</p>
<p>A totally ordered list of actions executed by the abstract machine is called an
<em>execution trace</em>. There might be several possible execution traces for a
given program due to non-determinism.</p>
<p>For a given execution trace, we define an irreflexive, transitive
<em>happens-before relation</em> that captures the causality between actions in
the OCaml program. The happens-before relation is defined as the smallest
transitive relation satisfying the following properties:</p>
<ul>
<li>
We define the order in which a domain executes its actions as the
<em>program order</em>. If an action <span><span>x</span></span> precedes another action <span><span>y</span></span> in
program order, then <span><span>x</span></span> precedes <span><span>y</span></span> in happens-before order.
</li>
<li>If <span><span>x</span></span> is a write to an atomic location and <span><span>y</span></span> is a subsequent read or
write to that memory location in the execution trace, then <span><span>x</span></span> precedes <span><span>y</span></span>
in happens-before order. For atomic locations, <span><span>compare_and_set</span></span>,
<span><span>fetch_and_add</span></span>, <span><span>exchange</span></span>, <span><span>incr</span></span> and <span><span>decr</span></span> are considered to perform
both a read and a write.
</li>
<li>If <span><span>x</span></span> is <span><span>Domain.spawn f</span></span> and <span><span>y</span></span> is the first action in the newly
spawned domain executing <span><span>f</span></span>, then <span><span>x</span></span> precedes <span><span>y</span></span> in happens-before
order.
</li>
<li>If <span><span>x</span></span> is the last action in a domain <span><span>d</span></span> and <span><span>y</span></span> is <span><span>Domain.join     d</span></span>, then <span><span>x</span></span> precedes <span><span>y</span></span> in happens-before order.
</li>
<li>If <span><span>x</span></span> is an unlock operation on a mutex, and <span><span>y</span></span> is any subsequent
operation on the mutex in the execution trace, then <span><span>x</span></span> precedes <span><span>y</span></span> in
happens-before order.
</li>
</ul>
<h3 id="s:datarace">
<span>2.3</span> Data race</h3>
<p>In a given trace, two actions are said to be <em>conflicting</em> if they access
the same non-atomic location, at least one is a write and neither is an
initialising write to that location.</p>
<p>We say that a program has a <em>data race</em> if there exists some execution
trace of the program with two conflicting actions and there does not exist a
happens-before relationship between the conflicting accesses. A program without
data races is said to be <em>correctly synchronised</em>.</p>
<h3 id="ss:drf_sc">
<span>2.4</span> DRF-SC</h3>
<p><span>DRF-SC guarantee:</span> A program without data races will only exhibit
sequentially consistent behaviours.</p>
<p>DRF-SC is a strong guarantee for the programmers. Programmers can use
<em>sequential reasoning</em> i.e., reasoning by executing one inter-domain
action after the other, to identify whether their program has a data race. In
particular, they do not need to reason about reorderings described in
section ‍<a href="https://ocaml.org/manual/5.2/memorymodel.html#s%3Awhy_relaxed_memory">10.1</a> in order to determine whether their program
has a data race. Once the determination that a particular program is data race
free is made, they do not need to worry about reorderings in their code.</p>
<h2 id="s:drf_reasoning">
<span>3</span> Reasoning with DRF-SC</h2>
<p>In this section, we will look at examples of using DRF-SC for program
reasoning. In this section, we will use the functions with names <span><span>dN</span></span> to
represent domains executing in parallel with other domains. That is, we assume
that there is a <span><span>main</span></span> function that runs the <span><span>dN</span></span> functions in parallel as
follows:</p>
<pre>let main () =
  let h1 = Domain.spawn d1 in
  let h2 = Domain.spawn d2 in
  ...
  ignore @@ Domain.join h1;
  ignore @@ Domain.join h2
</pre>
<p>
Here is a simple example with a data race:</p>
<div>

<div>

<p>
<span>let</span> r = <span>ref</span> 0
<span>let</span> d1 () = r := 1
<span>let</span> d2 () = !r</p>
</div>

</div>
<p><span><span>r</span></span> is a non-atomic reference. The two domains race to access the reference,
and <span><span>d1</span></span> is a write. Since there is no happens-before relationship between the
conflicting accesses, there is a data race.</p>
<p>Both of the programs that we had seen in the section ‍<a href="https://ocaml.org/manual/5.2/memorymodel.html#s%3Awhy_relaxed_memory">10.1</a>
have data races. It is no surprise that they exhibit non sequentially
consistent behaviours.</p>
<p>Accessing disjoint array indices and fields of a record in parallel is not a
data race. For example,</p>
<div>

<div>

<p>
<span>let</span> a = [| 0; 1 |]
<span>let</span> d1 () = a.(0) &lt;- 42
<span>let</span> d2 () = a.(1) &lt;- 42</p>
</div>

</div>
<div>

<div>

<p>
<span>type</span> t = {
  <span>mutable</span> a : int;
  <span>mutable</span> b : int
}
<span>let</span> r = {a = 0; b = 1}
<span>let</span> d1 () = r.a &lt;- 42
<span>let</span> d2 () = r.b &lt;- 42</p>
</div>

</div>
<p>do not have data races.</p>
<p>Races on atomic locations do not lead to a data race.</p>
<div>

<div>

<p>
<span>let</span> r = Atomic.make 0
<span>let</span> d1 () = Atomic.set r 1
<span>let</span> d2 () = Atomic.get r</p>
</div>

</div>
<h4 id="s:mm_msg_passing">Message-passing</h4>
<p>Atomic variables may be used for implementing non-blocking communication
between domains.</p>
<div>

<div>

<p>
<span>let</span> msg = <span>ref</span> 0
<span>let</span> flag = Atomic.make <span>false</span>
<span>let</span> d1 () =
  msg := 42; 
  Atomic.set flag <span>true</span> 
<span>let</span> d2 () =
  <span>if</span> Atomic.get flag  <span>then</span>
    !msg 
  <span>else</span> 0</p>
</div>

</div>
<p>Observe that the actions <span><span>a</span></span> and <span><span>d</span></span> write and read from the same non-atomic
location <span><span>msg</span></span>, respectively, and hence are conflicting. We need to establish
that <span><span>a</span></span> and <span><span>d</span></span> have a happens-before relationship in order to show that this
program does not have a data race.</p>
<p>The action <span><span>a</span></span> precedes <span><span>b</span></span> in program order, and hence, <span><span>a</span></span> happens-before
<span><span>b</span></span>. Similarly, <span><span>c</span></span> happens-before <span><span>d</span></span>. If <span><span>d2</span></span> observes the atomic variable
<span><span>flag</span></span> to be <span><span>true</span></span>, then <span><span>b</span></span> precedes <span><span>c</span></span> in happens-before order. Since
happens-before is transitive, the conflicting actions <span><span>a</span></span> and <span><span>d</span></span> are in
happens-before order. If <span><span>d2</span></span> observes the <span><span>flag</span></span> to be <span><span>false</span></span>, then the read
of <span><span>msg</span></span> is not done. Hence, there is no conflicting access in this execution
trace. Hence, the program does not have a data race.</p>
<p>The following modified version of the message passing program does have a data
race.</p>
<div>

<div>

<p>
<span>let</span> msg = <span>ref</span> 0
<span>let</span> flag = Atomic.make <span>false</span>
<span>let</span> d1 () =
  msg := 42; 
  Atomic.set flag <span>true</span> 
<span>let</span> d2 () =
  ignore (Atomic.get flag); 
  !msg </p>
</div>

</div>
<p>The domain <span><span>d2</span></span> now unconditionally reads the non-atomic reference <span><span>msg</span></span>.
Consider the execution trace:</p>
<pre>Atomic.get flag; (* c *)
!msg; (* d *)
msg := 42; (* a *)
Atomic.set flag true (* b *)
</pre>
<p>
In this trace, <span><span>d</span></span> and <span><span>a</span></span> are conflicting operations. But there is no
happens-before relationship between them. Hence, this program has a data race.</p>
<h2 id="s:local_drf">
<span>4</span> Local data race freedom</h2>
<p>The OCaml memory model offers strong guarantees even for programs with data
races. It offers what is known as <em>local data race freedom sequential
consistency (LDRF-SC)</em> guarantee. A formal definition of this property is beyond
the scope of this manual chapter. Interested readers are encouraged to read the
PLDI 2018 paper on <a href="https://doi.org/10.1145/3192366.3192421">Bounding Data
Races in Space and Time</a>.</p>
<p>Informally, LDRF-SC says that the data race free parts of the program remain
sequentially consistent. That is, even if the program has data races, those
parts of the program that are disjoint from the parts with data races are
amenable to sequential reasoning.</p>
<p>Consider the following snippet:</p>
<div>

<div>

<p><span>let</span> snippet () =
  <span>let</span> c = <span>ref</span> 0 <span>in</span>
  c := 42;
  <span>let</span> a = !c <span>in</span>
  (a, c)</p>
</div>

</div>
<p>Observe that <span><span>c</span></span> is a newly allocated reference. Can the read of <span><span>c</span></span> return a
value which is not 42? That is, can <span><span>a</span></span> ever be not <span><span>42</span></span>? Surprisingly, in the
C++ and Java memory models, the answer is yes. With the C++ memory model, if
the program has a data race, even in unrelated parts, then the semantics is
undefined. If this snippet were linked with a library that had a data race,
then, under the C++ memory model, the read may return any value. Since data
races on unrelated locations can affect program behaviour, we say that C++
memory model is not bounded in space.</p>
<p>Unlike C++, Java memory model is bounded in space. But Java memory model is not
bounded in time; data races in the future will affect the past behaviour. For
example, consider the translation of this example to Java. We assume a prior
definition of <span><span>Class c {int x;}</span></span> and a shared <em>non-volatile</em> variable <span><span>C g</span></span>. Now the snippet may be part of a larger program with parallel threads:</p>
<pre>(* Thread 1 *)
C c = new C();
c.x = 42;
a = c.x;
g = c;

(* Thread 2 *)
g.x = 7;
</pre>
<p>
The read of <span><span>c.x</span></span> and the write of <span><span>g</span></span> in the first thread are done on separate
memory locations. Hence, the Java memory model allows them to be reordered. As
a result, the write in the second thread may occur before the read of <span><span>c.x</span></span>,
and hence, <span><span>c.x</span></span> returns <span><span>7</span></span>.</p>
<p>The OCaml equivalent of the Java code above is:</p>
<div>

<div>

<p><span>let</span> g = <span>ref</span> None

<span>let</span> snippet () =
  <span>let</span> c = <span>ref</span> 0 <span>in</span>
  c := 42;
  <span>let</span> a = !c <span>in</span>
  (a, c)

<span>let</span> d1 () =
  <span>let</span> (a,c) = snippet () <span>in</span>
  g := Some c;
  a

<span>let</span> d2 () =
  <span>match</span> !g <span>with</span>
  | None -&gt; ()
  | Some c -&gt; c := 7</p>
</div>

</div>
<p>Observe that there is a data race on both <span><span>g</span></span> and <span><span>c</span></span>. Consider only the first
three instructions in <span><span>snippet</span></span>:</p>
<pre>let c = ref 0 in
c := 42;
let a = !c in
...
</pre>
<p>
The OCaml memory model is bounded both in space and time. The only memory
location here is <span><span>c</span></span>. Reasoning only about this snippet, there is neither the
data race in space (the race on <span><span>g</span></span>) nor in time (the future race on <span><span>c</span></span>).
Hence, the snippet will have sequentially consistent behaviour, and the value
returned by <span><span>!c</span></span> will be <span><span>42</span></span>.</p>
<p>The OCaml memory model guarantees that even for programs with data races,
memory safety is preserved. While programs with data races may observe
non-sequentially consistent behaviours, they will not crash.</p>
<h2 id="s:mm_semantics">
<span>5</span> An operational view of the memory model</h2>
<p>In this section, we describe the semantics of the OCaml memory model. A formal
definition of the operational view of the memory model is presented in section
3 of the PLDI 2018 paper on
<a href="https://doi.org/10.1145/3192366.3192421">Bounding Data Races in Space and
Time</a>. This section presents an informal description of the memory model with
the help of an example.</p>
<p>Given an OCaml program, which may possibly contain data races, the operational
semantics tells you the values that may be observed by the read of a memory
location. For simplicity, we restrict the intra-thread actions to just the
accesses to atomic and non-atomic locations, ignoring domain spawn and join
operations, and the operations on mutexes.</p>
<p>We describe the semantics of the OCaml memory model in a straightforward
small-step operational manner. That is, the semantics is described by an
abstract machine that executes one action at a time, arbitrarily picking one of
the available domains at each step. This is similar to the abstract machine
that we had used to describe the happens-before relationship in
section ‍<a href="https://ocaml.org/manual/5.2/memorymodel.html#s%3Ahappens_before">10.2.2</a>.</p>
<h3 id="ss:mm_non_atomic">
<span>5.1</span> Non-atomic locations</h3>
<p>In the semantics, we model non-atomic locations as finite maps from timestamps
<span><span>t</span></span> to values <span><span>v</span></span>. We take timestamps to be rational numbers. The timestamps
are totally ordered but dense; there is a timestamp between any two others.</p>
<p>For example,</p>
<pre>a: [t1 -&gt; 1; t2 -&gt; 2]
b: [t3 -&gt; 3; t4 -&gt; 4; t5 -&gt; 5]
c: [t6 -&gt; 5; t7 -&gt; 6; t8 -&gt; 7]
</pre>
<p>
represents three non-atomic locations <span><span>a</span></span>, <span><span>b</span></span> and <span><span>c</span></span> and their
histories. The location <span><span>a</span></span> has two writes at timestamps <span><span>t1</span></span> and <span><span>t2</span></span> with
values <span><span>1</span></span> and <span><span>2</span></span>, respectively. When we write <span><span>a: [t1 -&gt; 1; t2 -&gt; 2]</span></span>, we
assume that <span><span>t1 &lt; t2</span></span>. We assume that the locations are initialised with a
history that has a single entry at timestamp 0 that maps to the initial value.</p>
<h3 id="ss:mm_domains">
<span>5.2</span> Domains</h3>
<p>Each domain is equipped with a <em>frontier</em>, which is a map from non-atomic
locations to timestamps. Intuitively, each domain’s frontier records, for each
non-atomic location, the latest write known to the thread. More recent writes
may have occurred, but are not guaranteed to be visible.</p>
<p>For example,</p>
<pre>d1: [a -&gt; t1; b -&gt; t3; c -&gt; t7]
d2: [a -&gt; t1; b -&gt; t4; c -&gt; t7]
</pre>
<p>
represents two domains <span><span>d1</span></span> and <span><span>d2</span></span> and their frontiers.</p>
<h3 id="ss:mm_na_access">
<span>5.3</span> Non-atomic accesses</h3>
<p>Let us now define the semantics of non-atomic reads and writes. Suppose domain
<span><span>d1</span></span> performs the read of <span><span>b</span></span>. For non-atomic reads, the domains may read an
arbitrary element of the history for that location, as long as it is not older
than the timestamp in the domains’s frontier. In this case, since <span><span>d1</span></span> frontier
at <span><span>b</span></span> is at <span><span>t3</span></span>, the read may return the value <span><span>3</span></span>, <span><span>4</span></span> or <span><span>5</span></span>. A non-atomic
read does not change the frontier of the current domain.</p>
<p>Suppose domain <span><span>d2</span></span> writes the value <span><span>10</span></span> to <span><span>c</span></span> (<span><span>c := 10</span></span>). We pick a new
timestamp <span><span>t9</span></span> for this write such that it is later than <span><span>d2</span></span>’s frontier at
<span><span>c</span></span>. Note a subtlety here: this new timestamp might not be later than everything
else in the history, but merely later than any other write known to the writing
domain. Hence, <span><span>t9</span></span> may be inserted in <span><span>c</span></span>’s history either (a) between <span><span>t7</span></span>
and <span><span>t8</span></span> or (b) after <span><span>t8</span></span>. Let us pick the former option for our discussion.
Since the new write appears after all the writes known by the domain <span><span>d2</span></span> to
the location <span><span>c</span></span>, <span><span>d2</span></span>’s frontier at <span><span>c</span></span> is also updated. The new state of the
abstract machine is:</p>
<pre>(* Non-atomic locations *)
a: [t1 -&gt; 1; t2 -&gt; 2]
b: [t3 -&gt; 3; t4 -&gt; 4; t5 -&gt; 5]
c: [t6 -&gt; 5; t7 -&gt; 6; t9 -&gt; 10; t8 -&gt; 7] (* new write at t9 *)

(* Domains *)
d1: [a -&gt; t1; b -&gt; t3; c -&gt; t7]
d2: [a -&gt; t1; b -&gt; t4; c -&gt; t9] (* frontier updated at c *)
</pre>
<h3 id="ss:mm_at_access">
<span>5.4</span> Atomic accesses</h3>
<p>Atomic locations carry not only values but also synchronization information. We
model atomic locations as a pair of the value held by that location and a
frontier. The frontier models the synchronization information, which is merged
with the frontiers of threads that operate on the location. In this way,
non-atomic writes made by one thread can become known to another by
communicating via an atomic location.</p>
<p>For example,</p>
<pre>(* Atomic locations *)
A: 10, [a -&gt; t1; b -&gt; t5; c -&gt; t7]
B: 5,  [a -&gt; t2; b -&gt; t4; c -&gt; t6]
</pre>
<p>
shows two atomic variables <span><span>A</span></span> and <span><span>B</span></span> with values <span><span>10</span></span> and <span><span>5</span></span>,
respectively, and frontiers of their own. We use upper-case variable names to
indicate atomic locations.</p>
<p>During atomic reads, the frontier of the location is merged into that of the
domain performing the read. For example, suppose <span><span>d1</span></span> reads <span><span>B</span></span>. The read
returns <span><span>5</span></span>, and <span><span>d1</span></span>’s frontier updated by merging it with <span><span>B</span></span>’s frontier,
choosing the later timestamp for each location. The abstract machine state
before the atomic read is:</p>
<pre>(* Non-atomic locations *)
a: [t1 -&gt; 1; t2 -&gt; 2]
b: [t3 -&gt; 3; t4 -&gt; 4; t5 -&gt; 5]
c: [t6 -&gt; 5; t7 -&gt; 6; t9 -&gt; 10; t8 -&gt; 7]

(* Domains *)
d1: [a -&gt; t1; b -&gt; t3; c -&gt; t7]
d2: [a -&gt; t1; b -&gt; t4; c -&gt; t9]

(* Atomic locations *)
A: 10, [a -&gt; t1; b -&gt; t5; c -&gt; t7]
B: 5,  [a -&gt; t2; b -&gt; t4; c -&gt; t6]
</pre>
<p>
As a result of the atomic read, the abstract machine state is updated to:</p>
<pre>(* Non-atomic locations *)
a: [t1 -&gt; 1; t2 -&gt; 2]
b: [t3 -&gt; 3; t4 -&gt; 4; t5 -&gt; 5]
c: [t6 -&gt; 5; t7 -&gt; 6; t9 -&gt; 10; t8 -&gt; 7]

(* Domains *)
d1: [a -&gt; t2; b -&gt; t4; c -&gt; t7] (* frontier updated at a and b *)
d2: [a -&gt; t1; b -&gt; t4; c -&gt; t9]

(* Atomic locations *)
A: 10, [a -&gt; t1; b -&gt; t5; c -&gt; t7]
B: 5,  [a -&gt; t2; b -&gt; t4; c -&gt; t6]
</pre>
<p>
During atomic writes, the value held by the atomic location is updated. The
frontiers of both the writing domain and that of the location being written to
are updated to the merge of the two frontiers. For example, if <span><span>d2</span></span> writes <span><span>20</span></span>
to <span><span>A</span></span> in the current machine state, the machine state is updated to:</p>
<pre>(* Non-atomic locations *)
a: [t1 -&gt; 1; t2 -&gt; 2]
b: [t3 -&gt; 3; t4 -&gt; 4; t5 -&gt; 5]
c: [t6 -&gt; 5; t7 -&gt; 6; t9 -&gt; 10; t8 -&gt; 7]

(* Domains *)
d1: [a -&gt; t2; b -&gt; t4; c -&gt; t7]
d2: [a -&gt; t1; b -&gt; t5; c -&gt; t9] (* frontier updated at b *)

(* Atomic locations *)
A: 20, [a -&gt; t1; b -&gt; t5; c -&gt; t9] (* value updated. frontier updated at c. *)
B: 5,  [a -&gt; t2; b -&gt; t4; c -&gt; t6]
</pre>
<h3 id="s:mm_semantics_reasoning">
<span>5.5</span> Reasoning with the semantics</h3>
<p>Let us revisit an example from earlier (section <a href="https://ocaml.org/manual/5.2/memorymodel.html#s%3Awhy_relaxed_memory">10.1</a>).</p>
<div>

<div>

<p><span>let</span> a = <span>ref</span> 0
<span>and</span> b = <span>ref</span> 0

<span>let</span> d1 () =
  a := 1;
  !b

<span>let</span> d2 () =
  b := 1;
  !a

<span>let</span> main () =
  <span>let</span> h = Domain.spawn d2 <span>in</span>
  <span>let</span> r1 = d1 () <span>in</span>
  <span>let</span> r2 = Domain.join h <span>in</span>
  <span>assert</span> (not (r1 = 0 &amp;&amp; r2 = 0))</p>
</div>

</div>
<p>This program has a data race on <span><span>a</span></span> and <span><span>b</span></span>, and hence, the program may exhibit
non sequentially consistent behaviour. Let us use the semantics to show that
the program may exhibit <span><span>r1 = 0 &amp;&amp; r2 = 0</span></span>.</p>
<p>The initial state of the abstract machine is:</p>
<pre>(* Non-atomic locations *)
a: [t0 -&gt; 0]
b: [t1 -&gt; 0]

(* Domains *)
d1: [a -&gt; t0; b -&gt; t1]
d2: [a -&gt; t0; b -&gt; t1]
</pre>
<p>
There are several possible schedules for executing this program. Let us
consider the following schedule:</p>
<pre>1: a := 1 @ d1
2: b := 1 @ d2
3: !b     @ d1
4: !a     @ d2
</pre>
<p>
After the first action <span><span>a:=1</span></span> by <span><span>d1</span></span>, the machine state is:</p>
<pre>(* Non-atomic locations *)
a: [t0 -&gt; 0; t2 -&gt; 1] (* new write at t2 *)
b: [t1 -&gt; 0]

(* Domains *)
d1: [a -&gt; t2; b -&gt; t1] (* frontier updated at a *)
d2: [a -&gt; t0; b -&gt; t1]
</pre>
<p>
After the second action <span><span>b:=1</span></span> by <span><span>d2</span></span>, the machine state is:</p>
<pre>(* Non-atomic locations *)
a: [t0 -&gt; 0; t2 -&gt; 1]
b: [t1 -&gt; 0; t3 -&gt; 1] (* new write at t3 *)

(* Domains *)
d1: [a -&gt; t2; b -&gt; t1]
d2: [a -&gt; t0; b -&gt; t3] (* frontier updated at b *)
</pre>
<p>
Now, for the third action <span><span>!b</span></span> by <span><span>d1</span></span>, observe that <span><span>d1</span></span>’s frontier at <span><span>b</span></span>
is at <span><span>t1</span></span>. Hence, the read may return either <span><span>0</span></span> or <span><span>1</span></span>. Let us assume that it
returns <span><span>0</span></span>. The machine state is not updated by the non-atomic read.</p>
<p>Similarly, for the fourth action <span><span>!a</span></span> by <span><span>d2</span></span>, <span><span>d2</span></span>’s frontier at <span><span>a</span></span> is at
<span><span>t0</span></span>. Hence, this read may also return either <span><span>0</span></span> or <span><span>1</span></span>. Let us assume that it
returns <span><span>0</span></span>. Hence, the assertion in the original program, <span><span>assert (not (r1 = 0 &amp;&amp; r2 = 0))</span></span>, will fail for this particular execution.</p>
<h2 id="s:mm_tearing">
<span>6</span> Non-compliant operations</h2>
<p>There are certain operations which are not memory model compliant.</p>
<ul>
<li>
<span><span>Array.blit</span></span> function on float arrays may cause <em>tearing</em>. When an
unsynchronized blit operation runs concurrently with some overlapping write
to the fields of the same float array, the field may end up with bits from
either of the writes.
</li>
<li>With flat-float arrays or records with only float fields on 32-bit
architectures, getting or setting a field involves two separate memory
accesses. In the presence of data races, the user may observe tearing.
</li>
<li>The <span><span>Bytes</span></span> module ‍<a href="https://ocaml.org/manual/5.2/../5.2/api/Bytes.html"><span>Bytes</span></a> permits mixed-mode accesses
where reads and writes may be of different sizes. Unsynchronized mixed-mode
accesses lead to tearing.
</li>
</ul>
<hr>
<p>Copyright © 2024 Institut National de
Recherche en Informatique et en Automatique</p>
</div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via ocaml.org https://ocaml.org<br>
<br>
June 24, 2024 at 03:25PM
